File: Reference/Competitive Programmer's Handbook/BFS.cpp
#include <iostream>
#include <vector>
#include <map>
#include <tuple>
#include <queue>
using namespace std;

vector<int> adj[6];

void bfs(int x) {

    queue<int> q;
    int distance[6] = {0};
    bool visited[6] = {0};

    visited[x] = true;
    distance[x] = 0;
    q.push(x);
    while (!q.empty()) {
        int s = q.front(); q.pop();
        // process node s
        for (auto u : adj[s]) {
            if (visited[u]) continue;
            visited[u] = true;
            cout << "Visited node " << u << endl;
            distance[u] = distance[s]+1;
            cout << distance[u] << endl;
            q.push(u);
        }
    }

    for (int i = 1; i < 6; i++) {
        cout << "Node: " << i << " | Distance: " << distance[i] << endl;
    }
}

void undadd(int s, int r) {
    adj[s].push_back(r);
    adj[r].push_back(s);
}

int main() {

    undadd(1, 2);
    undadd(1, 4);
    undadd(2, 3);
    undadd(3, 5);
    undadd(2, 5);

    bfs(1);

    return 0;
}

File: CTCI/bfs.py

import Queue

""" nah too complicated

class Node:
	def __init__(self, cargo=None, adj=[]):
		self.cargo = cargo
		self.adj = adj

	def __str__(self):
		return str(self.cargo) + ' ' + str(self.adj)

	def __repr__(self):
		return str(self.cargo) + ' ' + str(self.adj)

# source: https://i.stack.imgur.com/2wxai.png

root = Node(5, [Node(11)])

print root

"""

# represent this graph through a dict: https://i.stack.imgur.com/2wxai.png

adj = {}

nodes = [2, 3, 5, 7, 8, 9, 10, 11]

# 2
# 3
map(lambda x : adj.setdefault(3, []).append(x), [10, 8])
# 5
adj[5] = [11]
# 7
adj[7]= [8, 11]
# 8
adj[8] = [9]
# 9
# 10
# 11
adj[11] = [2, 9, 10]

visited = {}

def bfs(root):
	q = Queue.Queue()
	visited[root] = True

	q.put(root)

	while not q.empty():
		this_node = q.get()
		print this_node
		visited[this_node] = True

		for adj_node in adj.setdefault(this_node, []):
			if not visited.setdefault(adj_node, False):
				q.put(adj_node)


bfs(7)